src/pybind/mgr/rook/rook_cluster.py

```diff
@@ -14,34 +15,174 @@
 
 # Optional kubernetes imports to enable MgrModule.can_run
 # to behave cleanly.
+from urllib3.exceptions import ProtocolError
+
+from mgr_util import merge_dicts
+
 try:
     from kubernetes.client.rest import ApiException
+    from kubernetes.client import V1ListMeta, CoreV1Api, V1Pod
+    from kubernetes import watch
 except ImportError:
     ApiException = None
 
+
+import orchestrator
+
+
 try:
-    import orchestrator
     from rook.module import RookEnv
-    from typing import List
+    from typing import List, Dict
 except ImportError:
     pass  # just used for type checking.
 
-
 log = logging.getLogger(__name__)
 
 
-class ApplyException(Exception):
+def _urllib3_supports_read_chunked():
+    # There is a bug in CentOS 7 as it ships a urllib3 which is lower
+    # than required by kubernetes-client
+    try:
+        from urllib3.response import HTTPResponse
+        return hasattr(HTTPResponse, 'read_chunked')
+    except ImportError:
+        return False
+
+
+_urllib3_supports_read_chunked = _urllib3_supports_read_chunked()
+
+class ApplyException(orchestrator.OrchestratorError):
     """
     For failures to update the Rook CRDs, usually indicating
     some kind of interference between our attempted update
     and other conflicting activity.
     """
 
 
+def threaded(f):
+    def wrapper(*args, **kwargs):
+        t = threading.Thread(target=f, args=args, kwargs=kwargs)
+        t.start()
+        return t
+
+    return wrapper
+
+
+class KubernetesResource(object):
+    def __init__(self, api_func, **kwargs):
+        """
+        Generic kubernetes Resource parent class
+
+        The api fetch and watch methods should be common across resource types,
+
+        Exceptions in the runner thread are propagated to the caller.
+
+        :param api_func: kubernetes client api function that is passed to the watcher
+        :param filter_func: signature: ``(Item) -> bool``.
+        """
+        self.kwargs = kwargs
+        self.api_func = api_func
+
+        # ``_items`` is accessed by different threads. I assume assignment is atomic.
+        self._items = dict()
+        self.thread = None  # type: threading.Thread
+        self.exception = None
+        if not _urllib3_supports_read_chunked:
+            logging.info('urllib3 is too old. Fallback to full fetches')
+
+    def _fetch(self):
+        """ Execute the requested api method as a one-off fetch"""
+        response = self.api_func(**self.kwargs)
+        # metadata is a V1ListMeta object type
+        metadata = response.metadata  # type: V1ListMeta
+        self._items = {item.metadata.name: item for item in response.items}
+        log.error('Full fetch of {}. result: {}'.format(self.api_func, len(self._items)))
```

Comment:
Should we low down the level here? This message is logged on every query on platform without watch support.

Reply:
fixed